### PR TITLE
implement setState

### PR DESCRIPTION
**Pull request by ghost**
Connects to #15

OK this doesn't implement everything in setState yet (e.g. doesn't do the callback param yet) but this is a start which should be good for most use cases.

setState doc: https://facebook.github.io/react/docs/react-component.html#setstate

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -14,23 +15,35 @@ abstract class NgComponent<Props, State> {
...
-  public $onChanges(changes: object) {
+  $onChanges(changes: object) {
     const oldProps = mapValues<{}, Props>(changes, 'previousValue')
     const newProps = mapValues<{}, Props>(changes, 'currentValue')
+    if (!this.__isFirstRender && !this.didPropsChange(newProps, oldProps)) return
+    this.lifecycle(newProps, {} as State)
+  }
 
+  // note: this isn't meant to exactly replicate React's setState (e.g. http://stackoverflow.com/q/28922275/7816712)
+  protected setState(newState: State) {
```
Comment (bcherny) [inline index.ts]: `Partial<State>`

File: index.ts [hunk 2]
```diff
@@ -14,23 +15,35 @@ abstract class NgComponent<Props, State> {
...
+    if (this.__setStateTriggersLifecycle) {
+      setTimeout(() => this.lifecycle({} as Props, newState), 0)
```
Comment (bcherny) [inline index.ts]: Is this how React does it?

File: test.ts [hunk 3]
```diff
@@ -258,6 +258,98 @@ describe('Component', () => {
     })
   })
 
+  describe('#setState', () => {
```
Comment (bcherny) [inline test.ts:261]: Can you add test cases for triggering setState in constructor, render(), shouldComponentUpdate(), componentWillUnmount()

File: index.ts [hunk 4]
```diff
@@ -14,23 +15,35 @@ abstract class NgComponent<Props, State> {
...
+      this.avoidSetStateLifecycle(() => this.componentWillMount())
       this.render()
       this.__isFirstRender = false
     } else {
-      if (!this.didPropsChange(newProps, oldProps)) return
-      this.componentWillReceiveProps(nextProps)
-      const shouldUpdate = this.shouldComponentUpdate(nextProps, this.state)
-      assign(this, { props: nextProps })
+      this.avoidSetStateLifecycle(() => this.componentWillReceiveProps(nextProps))
+      const shouldUpdate = this.shouldComponentUpdate(nextProps, nextState)
```
Comment (bcherny) [inline index.ts]: Can you add test for nextState in shouldComponentUpdate

File: index.ts [hunk 5]
```diff
@@ -14,23 +15,35 @@ abstract class NgComponent<Props, State> {
...
+      assign(this, { props: nextProps, state: nextState })
+      this.avoidSetStateLifecycle(() => this.componentWillMount())
       this.render()
       this.__isFirstRender = false
     } else {
-      if (!this.didPropsChange(newProps, oldProps)) return
-      this.componentWillReceiveProps(nextProps)
-      const shouldUpdate = this.shouldComponentUpdate(nextProps, this.state)
-      assign(this, { props: nextProps })
+      this.avoidSetStateLifecycle(() => this.componentWillReceiveProps(nextProps))
```
Comment (bcherny) [inline index.ts:45]: Can you add a test that `this.state` still refers to the old state inside componentWillReceiveProps

File: index.ts (hunk 2 above)
Comment (ghost) [reply index.ts]: set comment above - it's sometimes sync or async... haven't looked into exactly how this behavior works

**Comment by bcherny** [comment]
A few more behaviors, we can probably do in a separate PR/PRs:

- Callback API for setState (as you mentioned)
- Prevent direct assignment to state (at compile time, and at runtime)
- Implement forceUpdate()
- Use setImmediate instead of setTimeout (if the browser supports it)

Another thought: we can deviate from React here, and allow direct assignment to this.state (rather than force people to use setState) while still triggering appropriate lifecycle hooks if we use an ES6 proxy.

**Review by bcherny — APPROVED** [review]
Awesome work @khoomeister!

File: index.ts (hunk 2 above)
Comment (bcherny) [reply index.ts]: Looking at the SO post you linked, it seems to be always sync, no?

**Pull request merged**